// @hachther/mesomb/dist/operations/AOperation.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AOperation = void 0;
var MeSomb_1 = require("../MeSomb");
var Signature_1 = require("../Signature");
var exceptions_1 = require("../exceptions");
var AOperation = /** @class */ (function () {
    function AOperation(_a) {
        var target = _a.target, accessKey = _a.accessKey, secretKey = _a.secretKey, _b = _a.language, language = _b === void 0 ? 'en' : _b;
        this.target = target;
        this.accessKey = accessKey;
        this.secretKey = secretKey;
        this.language = language;
    }
    AOperation.prototype._buildUrl = function (endpoint) {
        return "".concat(MeSomb_1.default.HOST, "/api/").concat(MeSomb_1.default.APIVERSION, "/").concat(endpoint);
    };
    AOperation.prototype._getAuthorization = function (method, endpoint, date, nonce, headers, body) {
        if (headers === void 0) { headers = {}; }
        var url = this._buildUrl(endpoint);
        var credentials = { accessKey: this.accessKey, secretKey: this.secretKey };
        return Signature_1.Signature.signRequest(this.service, method, url, date, nonce, credentials, headers, body || {});
    };
    AOperation.prototype._processClientException = function (response) {
        return __awaiter(this, void 0, void 0, function () {
            var message, code, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, response.text()];
                    case 1:
                        message = _a.sent();
                        if (message.startsWith('{')) {
                            data = JSON.parse(message);
                            message = data.detail;
                            code = data.code;
                        }
                        switch (response.status) {
                            case 404:
                                throw new exceptions_1.ServiceNotFoundError(message);
                            case 403:
                            case 401:
                                throw new exceptions_1.PermissionDeniedError(message);
                            case 400:
                                throw new exceptions_1.InvalidClientRequestError(message, code);
                            default:
                                throw new exceptions_1.ServerError(message, code);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    AOperation.prototype.executeRequest = function (method, endpoint, date, nonce, body, mode) {
        if (body === void 0) { body = null; }
        if (mode === void 0) { mode = 'asynchronous'; }
        return __awaiter(this, void 0, void 0, function () {
            var url, headers, authorization, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = this._buildUrl(endpoint);
                        headers = {
                            'x-mesomb-date': String(date.getTime()),
                            'x-mesomb-nonce': nonce,
                            'Content-Type': 'application/json',
                            'X-MeSomb-OperationMode': mode,
                            'X-MeSomb-Source': "MeSombJS/v".concat(MeSomb_1.default.VERSION),
                            'Accept-Language': this.language,
                        };
                        if (this.service === 'payment') {
                            headers['X-MeSomb-Application'] = this.target;
                        }
                        if (this.service === 'fundraising') {
                            headers['X-MeSomb-Fund'] = this.target;
                        }
                        if (this.service === 'wallet') {
                            headers['X-MeSomb-Provider'] = this.target;
                        }
                        if (body === null || body === void 0 ? void 0 : body.trxID) {
                            headers['X-MeSomb-TrxID'] = String(body.trxID);
                            delete body.trxID;
                        }
                        if (method !== 'GET' && body) {
                            authorization = this._getAuthorization(method, endpoint, date, nonce, { 'content-type': 'application/json' }, body);
                        }
                        else {
                            authorization = this._getAuthorization(method, endpoint, date, nonce);
                        }
                        headers.Authorization = authorization;
                        return [4 /*yield*/, fetch(url, {
                                method: method,
                                body: body ? JSON.stringify(body) : undefined,
                                headers: headers,
                            })];
                    case 1:
                        response = _a.sent();
                        if (!(response.status >= 400)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this._processClientException(response)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [4 /*yield*/, response.json()];
                    case 4: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return AOperation;
}());
exports.AOperation = AOperation;
